### PR TITLE
Ports Families gamemode to Dynamic roundstart + midround rulesets

### LINKED ISSUE #49877: Families clutter traitor panel, all are uncategorized
This should probably be a nested option when setting leader/member status.
Also needs to be categorized.
image

### PR DESCRIPTION
**Pull request by tmtmtl30**
## About The Pull Request

Families is a decent gamemode. It isn't currently an option during dynamic rounds, unlike every other gamemode. Thus, it should be ported to dynamic. The current families code is not very conducive to this plan of action; even if it were, it would involve the duplication of hundreds of lines of code across 2-3 files that would need to be maintained in parallel in order to preserve parity between the dynamic ruleset and the standalone gamemode. This PR refactors the families code significantly in order to port it to dynamic (in the form of a roundstart ruleset and a rarer, shorter midround ruleset), done in two parts. Firstly, a significant amount of code is moved to the family member and space police antagonist datums. Secondly, the code that cannot be moved to the antagonist datums is instead moved to a new, unique "handler" datum. The families gamemode and ruleset instance this "handler" datum, feed it candidates, modify its variables directly after if necessary, and then call its procs. For example, the post_setup() and execute() procs of the families gamemode and dynamic rulesets respectively each call the "handler" proc post_setup_analogue(). This means that all the code is centralized and easier to maintain.

The mechanical impact on the standalone families gamemode is negligible. The families roundstart dynamic ruleset functions nearly identically to the standalone gamemode, barring slightly different criteria for eligibility. The families midround dynamic ruleset is very similar to the roundstart ruleset, though it is rarer and the space police will arrive 30-15 minutes after the ruleset's drafting as opposed to 60-30 minutes after the beginning of the round for the gamemode and roundstart ruleset. The roundstart dynamic ruleset and the midround dynamic ruleset cannot coexist; no round will have both.

Additionally, some bugs have been fixed as part of the refactor. The gamemode no longer iterates through all family members and cops to see if they're dead every time any person or slime dies, due to a misnamed proc. The space police 5 minute warning notification and arrival notification will occur at appropriate times instead of after the space police's arrival. An unrelated deprecated function and check have been removed. Finally, families are now nearly fully functional if they are added to players using the admin antagonist panel. They can invite others to their family and complete objectives, though there will be no space police, wanted levels, or point gain (outside of family members switching sides).

## Why It's Good For The Game

Families is an interesting gamemode, with a greater emphasis on roleplay and trust than any of our other antagonists. It's a shame that it can't pop up in dynamic, where the unique combinations of antagonists would really let it shine. This fixes that.

## Changelog
:cl:
fix: The Families gamemode and antagonist have been refactored and should perform more sanely in some cases.
add: Families can now appear in dynamic rounds.
/:cl:

Closes #49877
?????? #49935 -- whatever happened is still a mystery, but i think i stopped it from freezing everything and the code structure is so different now that whatever set of circumstances that came to pass to cause this issue might be no longer possible.

### CODE REVIEW

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 1]
```diff
@@ -208,6 +208,71 @@
...
+	name = "Family Head Aspirants"
+	persistent = TRUE
+	antag_flag = ROLE_FAMILIES
+	restricted_roles = list("Cyborg", "AI", "Prisoner","Security Officer", "Warden", "Detective", "Head of Security", "Captain", "Head of Personnel")
+	required_candidates = 2 // will always have at least 1 family + opponent for that family (other family or cop)
+	weight = 1
+	cost = 25
+	requirements = list(101,101,101,90,80,70,30,15,10,10) // more strict than the roundstart ruleset, because it's midround
+	high_population_requirement = 10
+	flags = TRAITOR_RULESET
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:227]:
```suggestion
```
TRAITOR_RULESET is for classic secret mode and shouldn't be used on a ruleset like this.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 2]
```diff
@@ -461,6 +461,53 @@
...
+	name = "Families"
+	persistent = TRUE
+	antag_flag = ROLE_FAMILIES
+	restricted_roles = list("Cyborg", "AI", "Prisoner","Security Officer", "Warden", "Detective", "Head of Security", "Captain", "Head of Personnel")
+	required_candidates = 2 // will always have at least 1 family + opponent for that family (other family or cop)
+	weight = 2
+	cost = 30
+	requirements = list(101,101,101,80,70,30,15,10,10,10)
+	high_population_requirement = 10
+	flags = TRAITOR_RULESET
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:480]:
```suggestion
```

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 3]
```diff
@@ -461,6 +461,53 @@
...
+	blocking_rules = list(/datum/dynamic_ruleset/midround/families)
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:481]: You got the blocking_rules the wrong way, this should go the midround ruleset with 
`/datum/dynamic_ruleset/roundstart/families`.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 4]
```diff
@@ -461,6 +461,53 @@
...
+	return handler.pre_setup_analogue()
+
+/datum/dynamic_ruleset/roundstart/families/execute()
+	return handler.post_setup_analogue(TRUE)
+
+/datum/dynamic_ruleset/roundstart/families/clean_up()
+	QDEL_NULL(handler)
+	..()
+
+/datum/dynamic_ruleset/roundstart/families/Destroy()
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:501]: rulesets generally don't get deleted, but if you want to you should make Destroy call clean_up on a lower level.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 5]
```diff
@@ -208,6 +208,71 @@
...
+	return handler.pre_setup_analogue()
+
+/datum/dynamic_ruleset/midround/families/execute()
+	return handler.post_setup_analogue(TRUE)
+
+/datum/dynamic_ruleset/midround/families/clean_up()
+	QDEL_NULL(handler)
+	..()
+
+/datum/dynamic_ruleset/midround/families/Destroy()
```
Comment (TheChosenEvilOne) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:267]: See the comment on /datum/dynamic_ruleset/roundstart/families/Destroy()

**Review by TheChosenEvilOne — COMMENTED** [review]
Just a quick review on the dynamic side of things.

**Comment by ATH1909** [comment]
Why does this PR have so many downthumbs?

**Comment by evilcyborgmrt** [comment]
Because nobody likes families

**Comment by tmtmtl30** [comment]
it's a matter of consistency. if families isn't well-designed, then it should be removed or reworked, not kept in a limbo state and shielded from our most populous server.

i had some ideas for switching up how induction packages are spawned + used, to prevent people from spamming them early in the round and bypassing the balance cap, but i didn't want to bikeshed even more than i already am. if you have mechanical suggestions to improve the mode (adjusting point gain, wanted levels, objectives for families / space police), i'd be happy to hear them.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm (hunk 3 above)
Comment (tmtmtl30) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:481]: that line is already there, i was just copying what the revs rulesets did.

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm (hunk 4 above)
Comment (tmtmtl30) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:501]: eh, i'll just remove the proc

**Comment by Iamgoofball** [comment]
I will be PRing the new families gangs to this PR to avoid merge conflicts

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm (hunk 3 above)
Comment (TheChosenEvilOne) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:480]: well, this just stops the roundstart ruleset from being triggered AFTER the midround has been triggered.

**Comment by Shadowflame909** [comment]
> Why does this PR have so many downthumbs?

because its just organized greytide with no mechanics.

every round is a families round as long as the greytide wears the same jumpsuit and gets a spray can from the autolathe.

**Comment by SkeletalElite** [comment]
keep families off bagil please and thank you

**Comment by Iamgoofball** [comment]
Stop posting garbage about whether you like or dislike families in this PR. I will be reporting all posts that do that.

**Comment by optimumtact** [comment]
Keep it about the specifics of families in dynamic please, if you want to complain about the gamemode we have a forum.

**Comment by BeeSting12** [comment]
How likely is this ruleset to be used?

**Review by kriskog — COMMENTED** [review]
Since families has pretty big round impact it should probably be treated as a highlander ruleset.
```
flags = HIGHLANDER_RULESET
```
This makes it a lot less likely to spawn with other highlanders, so nukies, cult, revs.

**Comment by tmtmtl30** [comment]
the roundstart ruleset has a weight of 2, the same as the roundstart revolutionary or roundstart wizard rulesets. the midround ruleset has a weight of 1, the same as the midround wizard ruleset. as these weights are relative to the others in their category (roundstart rulesets only compete with other roundstart rulesets and midround rulests only compete with other midround rulesets), they can't be compared to each other. 

by weight, the roundstart families ruleset composes 2/24ths or ~8.3% of the roundstart rulesets. this is the same as the revolution ruleset if the families ruleset exists, as they have the same weight -- for reference, if the families ruleset was not present, the revolution ruleset would make up 2/22nds (~9.1%) of the roundstart rulesets by weight. the midround families ruleset is significantly rarer, composing 1/31st or ~3.2% of the roundstart rulesets by weight. however, there are additional requirements for rulesets to be drafted -- they must meet a minimum amount of eligible candidates, minimum population, and minimum threat level dependent on the population. additionally, not all rounds will draft a midround ruleset, so it's hard to be more precise.

**Comment by kriskog** [comment]
Since dynamic will roll between 1 and 3 roundstart rulesets, the cost is another pretty important variable. At 25 cost it is effectively more likely to run than revs, as it will fit as secondary/tertiary ruleset more often.

**Comment by tmtmtl30** [comment]
yeah, that's true (though roundstart families is 30 cost, not 25)

**Comment by IndieanaJones** [comment]
How often does the midround variant of this show up?  Do I have to be concerned with families randomly showing up and ending a round due to the forced presence of Space Police?  Is there any way to stop it once the gamemode is triggered to prevent the Space Police?

Also, how is death calculated in a midround situation?  Do the Space Police factor in deaths on the station prior to the arrival of the gangs?  Does it count only crew, or do other carbons also get tallied?

**Comment by tmtmtl30** [comment]
midround families should be pretty rare, but it's hard for me to anchor it to something else since pretty much every other midround ruleset misunderstood the "high_population_requirement" variable, meaning that they're way less likely to spawn above 55 population (the variable represents the amount of threat needed to draft the rule, separate from the rule's _cost_, during a 55 or higher pop round, and the rulesets have incorrectly set it to 50, when the 48-53 pop range requirement is 10). i'll open another PR to fix that. it's maybe a lot less likely than midround nukies, though, and i've never thought of those as particularly common. on populations in the 48-54 range, it's slightly less likely than midround wizard, and <1/4th as likely as a blob or a space dragon.

currently, the midround ruleset would count deaths from the start of the shift, which does seem a little bit silly. i'll fix that too.

**Comment by tmtmtl30** [comment]
on second thought, i'm not too sure. if there are a lot of deaths before the midround ruleset activates, then presumably things aren't going great on-station and leaving quickly is probably in order _and_ a large force of highly-equipped cops would make sense; _then_ you run into the problem of cucking antagonists by bringing in a bunch of highly-equipped soldiers. i'm curious what other people think about this

**Comment by tmtmtl30** [comment]
okay, i've added a check such that if the deaths_during_shift is greater than half of the people who were readied up at roundstart, then the families midround ruleset will fail. this is in addition to the requirement of 36 living players.

**Comment by nightred** [comment]
I believe that the deaths count related to bringing in the cops should only count after the game mode has been activated, and should have a way of ending without ending the round completely.

**Comment by tmtmtl30** [comment]
that's the way i have deathcounting handled. in families, they don't end the round, just delay the shuttle's departure until 10 minutes after the cops arrive. i could probably change that too.

**Comment by tmtmtl30** [comment]
reeeeee

**Comment by kriskog** [comment]
Ill look at this soonish

**Review by kriskog — COMMENTED** [review]
Merge conflict, otherwise seems merge ready.

**Comment by tmtmtl30** [comment]
should be fixed

**Comment by kriskog** [comment]
We might need to tweak stuff, like the chance of occurence. Lets see how it works out

**Pull request merged**